### PR TITLE
Parenthesize unsplittable nodes if it makes them fit

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR implements black's logic where it parenthesizes unsplittable nodes like constants or identifiers
if it makes them fit the line

closes https://github.com/astral-sh/ruff/issues/6271

## Test Plan

I added new test plans, diffed the output for `django` to ensure this PR introduces no regression, and verified that the similarity indices improve

**main**

| project      | similarity index |
|--------------|------------------|
| cpython      | 0.75473          |
| django       | 0.99805          |
| transformers | 0.99620          |
| twine        | 0.99876          |
| typeshed     | 0.99953          |
| warehouse    | 0.99601          |
| zulip        | 0.99727          |

**This PR**

| project      | similarity index |
|--------------|------------------|
| cpython      | 0.75475          |
| django       | 0.99880          |
| transformers | 0.99629          |
| twine        | 0.99929          |
| typeshed     | 0.99953          |
| warehouse    | 0.99613          |
| zulip        | 0.99852          |

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6734** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6734" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/6734?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/expression/mod.rs [hunk 1]
```diff
@@ -223,6 +223,35 @@ impl Format<PyFormatContext<'_>> for MaybeParenthesizeExpression<'_> {
                     }
                 }
             },
+            OptionalParentheses::IfFits => match parenthesize {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/mod.rs:226]: I'm not very happy with this name. The branches are the same as for `Never` except for `IfBreaks`.

File: crates/ruff_formatter/src/builders.rs
Comment (MichaReiser) [inline crates/ruff_formatter/src/builders.rs:1]: The formatter changes bring back the `BestFitting::print_mode`

File: crates/ruff_formatter/src/printer/mod.rs [hunk 2]
```diff
@@ -311,6 +314,14 @@ impl<'a> Printer<'a> {
                     _ => {
                         self.state.measured_group_fits = true;
 
+                        // Set the print mode for this group in case this group was measured as part of a
+                        // previous `fits_group` call.
+                        if let Some(id) = id {
```
Comment (MichaReiser) [inline crates/ruff_formatter/src/printer/mod.rs:319]: This was a bug in the Printer where group modes set during a previous `fits` test stuck around.

**Comment by MichaReiser** [comment]
Ouch, this is, not too surprisingly, not good for performance.

**Comment by charliermarsh** [comment]
Is that mostly due to the `best_fitting!` or is there something else going on?

**Comment by MichaReiser** [comment]
It's a combination of memoized and best fitting, both of them allocate. 

The solution is to limit the condition of when to use this layout. It should only be necessary if the content length is close to the max line length (I have to do some math tomorrow)

**Comment by MichaReiser** [comment]
Splitting the PR must have confused graphite. See https://github.com/astral-sh/ruff/pull/6816

**Pull request closed without merging**